### ISSUE TITLE
Adapt flexible -newline- and -space- rules to use nonwhitespace approach

### ISSUE THREAD

**Issue by m-allanson** [stylelint/stylelint#205]
## Value list

* [ ] `value-list-comma-newline-after`: `"always"|"always-multi-line"`
* [ ] `value-list-comma-newline-before`: `"always"|"always-multi-line"`
* [ ] `value-list-comma-space-after`: `"always"|"always-single-line"`
* [ ] `value-list-comma-space-before`: `"always"|"always-single-line"`
* [ ] `value-list-comma-nonwhitespace-before`: `"always"|"always-multi-line"|"always-single-line"`
* [ ] `value-list-comma-nonwhitespace-after`: `"always"|"always-multi-line"|"always-single-line"`

## Closing brace

* [ ] `block-closing-brace-newline-after`: `"always"|"always-multi-line"`
* [ ] `block-closing-brace-newline-before`: `"always"|"always-multi-line"`
* [ ] `block-closing-brace-space-after`: `"always"|"always-single-line"`
* [ ] `block-closing-brace-space-before`: `"always"|"always-single-line"`
* [ ] `block-closing-brace-nonwhitespace-before`: `"always"|"always-multi-line"|"always-single-line"`
* [ ] `block-closing-brace-nonwhitespace-after`: `"always"|"always-multi-line"|"always-single-line"`

## Opening brace

* [ ] `block-opening-brace-newline-after`: `"always"|"always-multi-line"`
* [ ] `block-opening-brace-newline-before`: `"always"|"always-multi-line"`
* [ ] `block-opening-brace-space-after`: `"always"|"always-single-line"`
* [ ] `block-opening-brace-space-before`: `"always"|"always-single-line"`
* [ ] `block-opening-brace-nonwhitespace-before`: `"always"|"always-multi-line"|"always-single-line"`
* [ ] `block-opening-brace-nonwhitespace-after`: `"always"|"always-multi-line"|"always-single-line"`

Some of the rules related to newlines reference "a single newline" for their `always` options and "any whitespace" for their `never` options.  Is this correct?  

Some examples are:

- https://github.com/stylelint/stylelint/blob/master/src/rules/rule-semicolon-newline-after/README.md
- https://github.com/stylelint/stylelint/blob/master/src/rules/block-opening-brace-newline-after/README.md
- https://github.com/stylelint/stylelint/blob/master/src/rules/value-list-comma-newline-before/README.md

I'd be happy to do a PR to update these, if required.

**Comment by davidtheclark** [comment]
Yes, that's correct. It's a design point I don't think any of us are totally certain about. But our thinking was that, for example, with the rule `block-opening-brace-newline-after` a user would want one of the following conventions:

1. Have a newline after opening braces
2. Have some other whitespace character after opening braces
3. Have no whitespace after opening braces
4. Not care what's after opening braces

1 is addressed by using the rule with `"always"`.
2 is addressed by using `block-opening-brace-space-after` instead, and not using this rule.
3 is addressed by using the rule with `"never"`.
4 is addressed by not using this rule.

I'm not sure there's a use case for "Have any character except a newline after opening braces." Does that make sense?

With that explanation, do you think we're on the right track or do you think there's a better way?

**Comment by davidtheclark** [comment]
One alternative approach I can think of would be to leave off the `"never"` option for this and similar rules, and instead create a rule like `block-opening-brace-no-whitespace-after`.

**Comment by m-allanson** [comment]
Thanks, yes that does make sense.  Do you think there's a case for documenting patterns like this?  I assume [rule-guidelines.md](https://github.com/stylelint/stylelint/blob/master/docs/rule-guidelines.md) is aimed at stylelint developers, maybe there could be an equivalent file aimed at stylelint users.

Your alternative approach has the benefit of ensuring that rule actions are more accurately aligned with rule names.  Are there any downsides to that approach?  I haven't spent enough time with the project to have a strong opinion either way, sorry!

I think what initially confused me is the example for the `"never"` option in `rule-semicolon-newline-after` looks to be incorrect.  In [these lines](https://github.com/stylelint/stylelint/blame/master/src/rules/rule-semicolon-newline-after/README.md#L62-#L64), given the `"never"` option, then whitespace after the first semi-colon _should_ be considered a warning?

(Incidentally, I started looking at `rule-semicolon-newline-after` to see if there was support for #208.  As this was my main cause of warnings when trying stylelint on an existing project. Generally it's working really nicely though.)

**Comment by davidtheclark** [comment]
As I think about this confusion I think I'm leaning a little towards that alternative approach of `block-opening-brace-no-whitespace-after`. Any thoughts @stylelint/owners ?

**Comment by jeddy3** [comment]
> I think what initially confused me is the example for the "never" option in rule-semicolon-newline-after looks to be incorrect. In these lines, given the "never" option, then whitespace after the first semi-colon should be considered a warning?

Yeap. That example is wrong. My bad!

> Thanks, yes that does make sense. Do you think there's a case for documenting patterns like this? I assume rule-guidelines.md is aimed at stylelint developers, maybe there could be an equivalent file aimed at stylelint users.

Yes. Perhaps we can stick a stripped-down version of the rule-guidelines.md before the list of rules in rules.md. Or, perhaps the relevant bits of documentation should sit in each of the individual README.md files. This latter approach fits well with our "isolated" rules approach, but there could be a fair bit of duplication and documentation overhead.

> As I think about this confusion I think I'm leaning a little towards that alternative approach of block-opening-brace-no-whitespace-after. 

This sounds good to me. It did feel a bit off that the `"never"` options for both the `-*newline-* and `*-space-*` rules enforced the same thing. Shall we think this through (as a there's a fair few whitespace rules that will be affected by our decision here)?

Given the following rules:

* `value-list-comma-newline-after`: `"always"|"always-multi-line"`
* `value-list-comma-newline-before`: `"always"|"always-multi-line"`
* `value-list-comma-space-after`: `"always"|"always-single-line"`
* `value-list-comma-space-before`: `"always"|"always-single-line"`
* `value-list-comma-no-whitespace-before`
* `value-list-comma-no-whitespace-after`

Can anyone think of a style of that can't be enforced?

e.g. With:

```js
value-list-comma-newline-after: [2, "always-multi-line"]
value-list-comma-space-after: [2, "always-single-line"]
value-list-comma-no-whitespace-before: 2
```

You'd be able to enforce these two types of value lists in the same CSS document:

```css
a {
  font-family: sans, serif, monospace /* value-list single-line with space after */
}

b {
  box-shadow: 1px 1px 1px #000, /* value-list multi-line with newline after... */
    2px 2px 1px 1px #ccc inset,     /* ... */
    2px 2px 1px 2px #ccc inset;     /* ... */
}
```

Or...

```js
value-list-comma-space-before: [2, "always"]
value-list-comma-no-whitespace-after: 2
``` 

To enforce always a single line with space before the coma

```css
a {
  font-family: sans ,serif ,monospace
}
```

Or...

```js
value-list-comma-newline-before: [2, "always-multi-line"]
value-list-comma-space-after: [2, "always"]
```

To enforce this pattern that @MoOx has seen in the wild:

```css
a {
  font-family: sans, serif, monospace
}

b {
  box-shadow: 1px 1px 1px #000
    , 2px 2px 1px 1px #ccc inset
    , 2px 2px 1px 2px #ccc inset;
}
```

Oh no, I can't see a way of enforcing no whitespace before the comma only if it's a single line. Are we asking for too much flexibility here? It is possible if the `*-no-whitespace-*` rules except options e.g.

* `value-list-comma-nothing-before`: `"always"|"always-multi-line"|"always-single-line"`
* `value-list-comma-nothing-after`: `"always"|"always-multi-line"|"always-single-line"`

Then you could enforce the last CSS example with:

```js
value-list-comma-newline-before: [2, "always-multi-line"]
value-list-comma-space-after: [2, "always"]
value-list-comma-nothing-before: [2, "always-single-line"]
```

Thoughts?

Thanks for flagging this up btw @m-allanson !

**Comment by davidtheclark** [comment]
I'm fine with this proposal to have special rules for enforcing no whitespace around things.

`nothing-before` seems like a misnomer to me, though, as there is *something* before. I know `value-comma-no-whitespace-before` would violate the typical `*-no-*` convention ... We could tweak it just a little bit into `value-comma-nonwhitespace-before`? What do you think?

**Comment by jeddy3** [comment]
> I'm fine with this proposal to have special rules for enforcing no whitespace around things.

Groovy.

> nothing-before seems like a misnomer to me, though, as there is something before. I know value-comma-no-whitespace-before would violate the typical *-no-* convention ... We could tweak it just a little bit into value-comma-nonwhitespace-before? What do you think?

II thought you'd pick up on that! I felt the same way when I wrote it :) Not sure about `nonwhitespace` though as I think of a digit or a character as non-whitespace. 

I can think of a couple of options:

1. Think of another word.
2. Question the importance of the typical `*-no-*` convention.
3. As in #207 , we could use an optional option object for the `value-list-comma-no-whitespace-before` e.g. `value-list-comma-no-whitespace-before: [2, { ignore: ["multi-line"] }]`. Or is it a bit odd to have `"multi-line"` and `"single-line"` in an optional option object when they are first-case options for the other `value-list-comma-*-before` rules?
4. Swap the options from `"always"` to `"never"` e.g. `value-list-comma-whitespace-before`: `"never"|"never-multi-line"|"never-single-line"`

Any of those work?

**Comment by davidtheclark** [comment]
I don't understand what's wrong with `nonwhitespace`. I also think of a digit or character as non-whitespace --- isn't that we're saying: anything that isn't whitespace?

Buuuut, I'd be just as fine loosening up on the `*-no-*` convention and having `value-list-comma-no-whitespace-before`.

Either way, I think we'd go with `"always"|"always-multi-line"|"always-single-line"` as options. That seems like the clearest solution in my mind. @m-allanson does that seem clear to you? Or @MoOx?

(Re 4: I much prefer `value-list-comma-no-whitespace-before` to `value-list-comma-whitespace-before` with `"never"` options, because I think it's less confusing given the other related rules.)

**Comment by jeddy3** [comment]
> I don't understand what's wrong with nonwhitespace. I also think of a digit or character as non-whitespace --- isn't that we're saying: anything that isn't whitespace?

Whoops, my bad. I see what you're saying now and, yes, does that makes sense! :)

```js
value-list-comma-nonwhitespace-before: [2, "always"]
```

```css
a { 
  font-family: sans, serif, monospace;
/**               ↑      ↑
*  Is referring to the "s" and "f" characters */
```

So, I'm down with:

* `value-list-comma-nonwhitespace-before`: `"always"|"always-multi-line"|"always-single-line"`
* `value-list-comma-nonwhitespace-after`: `"always"|"always-multi-line"|"always-single-line"`

It also lets us keep the `*-no-*` convention for now, which I think is useful until we have a config checker in place.

@m-allanson Sound good to you? i.e. given:

* `value-list-comma-newline-after`: `"always"|"always-multi-line"`
* `value-list-comma-newline-before`: `"always"|"always-multi-line"`
* `value-list-comma-space-after`: `"always"|"always-single-line"`
* `value-list-comma-space-before`: `"always"|"always-single-line"`
* `value-list-comma-nonwhitespace-before`: `"always"|"always-multi-line"|"always-single-line"`
* `value-list-comma-nonwhitespace-after`: `"always"|"always-multi-line"|"always-single-line"`

Can you enforce the styleguide you use?

I'll also add something to the docs to explain the difference between `single-line` and `multi-line` value-lists e.g.

`single-line` value-list:

```css
a {
  color: pink;
  font-family: sans, serif, monospace; /* values all on one line */
}
```

`multi-line` value-list:

```css
a {
  box-shadow: 1px 1px 1px #000 /* values across... */ 
    , 2px 2px 1px 1px #ccc inset   /* ... multiple ... */
    , 2px 2px 1px 2px #ccc inset;  /* ... lines */
  color: pink;
}
```

Sound good?

**Comment by davidtheclark** [comment]
SGTM

**Comment by m-allanson** [comment]
This all looks great to me. 

One more question, does the top comment for this issue also need to refer to some of the other `newline` and `space` rules?  The majority of these would become plain on / off rules:

## Media query list

* [ ] `media-query-list-comma-newline-after`
* [ ] `media-query-list-comma-newline-before`
* [ ] `media-query-list-comma-nonwhitespace-after`
* [ ] `media-query-list-comma-nonwhitespace-before`

## Selector delimiter

* [ ] `selector-delimiter-newline-after`
* [ ] `selector-delimiter-newline-before`
* [ ] `selector-delimiter-nonwhitespace-after`
* [ ] `selector-delimiter-nonwhitespace-before`

## Rule semicolon

* [ ] `rule-semicolon-newline-after`: `"always"|"always-multi-line"`
* [ ] `rule-semicolon-newline-before`: `"always"|"always-multi-line"`
* [ ] `rule-semicolon-space-after`: `"always"|"always-single-line"`
* [ ] `rule-semicolon-space-before`: `"always"|"always-single-line"`
* [ ] `rule-semicolon-nonwhitespace-before`
* [ ] `rule-semicolon-nonwhitespace-after`

**Comment by jeddy3** [comment]
@davidtheclark If you'd planned to start on this, then don't just yet. There's a lot of work here and I'm not 100% sure this is right route to go down. Let me sleep on it.

**Comment by davidtheclark** [comment]
Sounds good. Also I will look around a little to see if other linters address this well.

**Comment by davidtheclark** [comment]
One thing I just noticed ... if we do use `nonwhitespace` we should add a hyphen, `non-whitespace`, to be consistent with our other "non"s. (I'm thinking of `*-non-nested-*`)

**Comment by jeddy3** [comment]
Ok guys. I'm going to do a fair bit of back-peddling here, so please bare with me. I think the fact that each of the `-newline-` rules also has a `"never"` option totally confused me and set us off down a path we didn't need to go down. I think it's worth the extra discussion as I think we're so close to getting this one right.

In fact, I think what we have already works really well.

It works great for `before` and `after` rules where it doesn't make sense to also support newlines:

* `declaration-bang-space-after`: `"always"|"never"`
* `declaration-bang-space-before`: `"always"|"never"`
* `declaration-colon-space-after`: `"always"|"never"`
* `declaration-colon-space-before`: `"always"|"never"`
* `function-comma-space-after`: `"always"|"never"`
* `function-comma-space-before`: `"always"|"never"`
* `function-space-after`: `"always"|"never"`
* `media-feature-colon-space-after`: `"always"|"never"`
* `media-feature-colon-space-before`: `"always"|"never"`
* `media-feature-range-operator-space-after`: `"always"|"never"`
* `media-feature-range-operator-space-before`: `"always"|"never"`
* `selector-combinator-space-after`: `"always"|"never"`
* `selector-combinator-space-before`: `"always"|"never"`

And it also works great for the `inside` rules:

* `function-parentheses-space-inside`: `"always"|"never"`
* `media-query-parentheses-space-inside`: `"always"|"never"`
* `comment-space-inside`: `"always"|"never"`

It also aligns well with eslint's use of `space-*` with `"always"|"never"` options for its `before|after` and its `in` rules:

* http://eslint.org/docs/rules/space-after-keywords
* http://eslint.org/docs/rules/space-in-parens

It's a little odd (although arguably it *is* intuitive) that `space: "always"` means a single space and `space: "never"` means no whitespace, but I think it's better than the bloat of having to add a corresponding `-nonwhitespace-` to every `-before`, `-after` and `-inside` rule. The alternative is only to only add `-nonwhitespace-` rules for each of the `newline` rules, but that would create a big inconsistency in how the options work across the `-space-` rules.

We can already enforce the three previous examples i.e. given (think of the `-newline-` rules as complimentary to the `space` ones):

* `value-list-comma-space-after`: `"always"|"never"|"always-single-line"|"never-single-line"|"always-multi-line"|"never-multi-line"}`
* `value-list-comma-space-before`: `"always"|"never"|"always-single-line"|"never-single-line"|"always-multi-line"|"never-multi-line"}`
* `value-list-comma-newline-after`: `"always"|"always-multi-line"`
* `value-list-comma-newline-before`: `"always"|"always-multi-line"`

With:

```js
value-list-comma-newline-after: [2, "always-multi-line"]
value-list-comma-space-after: [2, "always-single-line"]
value-list-comma-space-before: [2, "never']
```

We can already enforce these two types of value lists in the same CSS document:

```css
a {
  font-family: sans, serif, monospace /* value-list single-line with space after */
}

b {
  box-shadow: 1px 1px 1px #000, /* value-list multi-line with newline after... */
    2px 2px 1px 1px #ccc inset,     /* ... */
    2px 2px 1px 2px #ccc inset;     /* ... */
}
```

Or...

```js
value-list-comma-space-before: [2, "always"]
value-list-comma-space-after: [2, "never"]
``` 

We can already always enforce a single line with space before the coma

```css
a {
  font-family: sans ,serif ,monospace
}
```

Or...

```js
value-list-comma-newline-before: [2, "always-multi-line"]
value-list-comma-space-after: [2, "always"]
value-list-comma-space-before: [2, "never-single-line"]
```

To enforce this pattern:

```css
a {
  font-family: sans, serif, monospace
}

b {
  box-shadow: 1px 1px 1px #000
    , 2px 2px 1px 1px #ccc inset
    , 2px 2px 1px 2px #ccc inset;
}
```

So, I suggest we simply strip the `"never-*"` options from the `-newline-` rules and all it a day :)

Am I missing something or does this look alright to you guys too? I guess we had to go around the houses a bit to get back to the start.

**Comment by davidtheclark** [comment]
I think that's fair. I guess I'm kind of wishy-washy on this: I keep saying yes.

We'd just have to make sure that documentation for `-newline-` rules pointed to the corresponding `-space-` rule as *the* way to enforce no whitespace.

**Comment by jeddy3** [comment]
>  I guess I'm kind of wishy-washy on this: I keep saying yes.

Ha, I think we both are! :) I think it's because we looked at this problem in isolation and found a good solution. But it wasn't until I went to implement that solution did I realise it doesn't really work in the context of the whole system.

Anyway, I think this exercise has really clarified (for me at least) how all the whitespace rules work together :)

I also realised this morning that the set of options I posted last night weren't quite right. It should be this:

* `value-list-comma-space-after`: `"always"|"never"|"always-single-line"|"never-single-line"`
* `value-list-comma-space-before`: `"always"|"never"|"always-single-line"|"never-single-line"`
* `value-list-comma-newline-after`: `"always"|"always-multi-line|"never-multi-line"`
* `value-list-comma-newline-before`: `"always"|"always-multi-line"|"never-multi-line"`

I need to look through all the whitespace rules to make sure they all consistently support these sets of options. I think they do. The only change should be removing the plain `"never"` option from the `newline` rules.

> We'd just have to make sure that documentation for -newline- rules pointed to the corresponding -space- rule as the way to enforce no whitespace.

Yeap. Will do. I've also written up a user guide (in upcoming PR) to explain all this whitespace stuff :)
